scipy/ndimage/src/ni_iterator_utils.h

```diff
@@ -0,0 +1,393 @@
+        line += line_stride;                                    \
+    }                                                           \
+}
+
+WRITE_LINE_FUNC(npy_bool)
+WRITE_LINE_FUNC(npy_ubyte)
+WRITE_LINE_FUNC(npy_ushort)
+WRITE_LINE_FUNC(npy_uint)
+WRITE_LINE_FUNC(npy_ulong)
+WRITE_LINE_FUNC(npy_ulonglong)
+WRITE_LINE_FUNC(npy_byte)
+WRITE_LINE_FUNC(npy_short)
+WRITE_LINE_FUNC(npy_int)
+WRITE_LINE_FUNC(npy_long)
+WRITE_LINE_FUNC(npy_longlong)
+WRITE_LINE_FUNC(npy_float)
+WRITE_LINE_FUNC(npy_double)
+
+
+static NPY_INLINE write_line_func*
+_get_write_line_func(enum NPY_TYPES typenum)
+{
+    /* These functions must be in the same order as enum NPY_TYPES. */
+    write_line_func *_lbi_write_funcs[] = {
+        &_write_line_npy_bool,
+        &_write_line_npy_byte, &_write_line_npy_ubyte,
+        &_write_line_npy_short, &_write_line_npy_ushort,
+        &_write_line_npy_int, &_write_line_npy_uint,
+        &_write_line_npy_long, &_write_line_npy_ulong,
+        &_write_line_npy_longlong, &_write_line_npy_ulonglong,
+        &_write_line_npy_float, &_write_line_npy_double,
+    };
+    if (_is_lbi_supported_type(typenum)) {
+        return _lbi_write_funcs[typenum];
+    }
+    return NULL;
+}
+
+
+/*
+ ***********************************************************************
+ ***                    Helpers for LBI_ExtendLine                   ***
+ ***********************************************************************
+ */
+
+typedef void (extend_line_func)(npy_double*, npy_intp, npy_intp, npy_intp,
+                                npy_double);
+
+static void
+_extend_line_nearest(npy_double *buffer, npy_intp line_length,
+                     npy_intp size_before, npy_intp size_after,
+                     npy_double unused_value)
+{
+    /* aaaaaaaa|abcd|dddddddd */
+    npy_double * const first = buffer + size_before;
+    npy_double * const last = first + line_length;
+    npy_double *dst, val;
+
+    val = *first;
+    dst = buffer;
+    while (size_before--) {
+        *dst++ = val;
+    }
+    dst = last;
+    val = *(last - 1);
+    while (size_after--) {
+        *dst++ = val;
+    }
+}
+
+static void
+_extend_line_wrap(npy_double *buffer, npy_intp line_length,
+                  npy_intp size_before, npy_intp size_after,
+                  npy_double unused_value)
+{
+    /* abcdabcd|abcd|abcdabcd */
+    npy_double * const first = buffer + size_before;
+    npy_double * const last = first + line_length;
+    const npy_double *src;
+    npy_double *dst;
+
+    src = last - 1;
+    dst = first - 1;
+    while (size_before--) {
+        *dst-- = *src--;
+    }
+    src = first;
+    dst = last;
+    while (size_after--) {
+        *dst++ = *src++;
+    }
+}
+
+static void
+_extend_line_reflect(npy_double *buffer, npy_intp line_length,
+                     npy_intp size_before, npy_intp size_after,
+                     npy_double unused_value)
+{
+    /* abcddcba|abcd|dcbaabcd */
```

Comment:
This comment and the one in the function below are the same.

I remember there were issues with one of these, but not with which one.